FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,95 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let managedCache = try ManagedCache.newUniqueInstance(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedFeedImage.images(from: feed, in: context)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.find(in: context).map(context.delete).map(context.save)
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = self.context
+		context.perform { action(context) }
+	}
+}
+
+@objc(ManagedCache)
+internal class ManagedCache: NSManagedObject {
```

Comment:
`internal` is the default access level, so you don't need to write it explicitly.

It was written explicitly in the first lectures of the program for teaching purposes, but you don't need to do it in your code.

Also this class is not intended to be subclassed thus it should be final

Please, remove all explicit `internal` keywords and make the class final.
```suggestion
final class ManagedCache: NSManagedObject {
```